### PR TITLE
Add of the TicTacToe contract

### PR DESCRIPTION
**Pull request by Rubyt0x**
# Summary

1. First version of the TicTacToe contract

2.  Functions implemented:
- New game
- Make moves
- Next player
- Save moves
- Get player positions
- Calculate winner
- Get winner
- Save winner
- Check alignments (horizontal, vertical, diagonal)
- Check if the game board is full

3. Issues:
- Bugs in the alignment functions

#TODO
- Bug fixes
- Tests
- Start the front

### CODE REVIEW

File: Contracts/TicTacToe/src/main.sw [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+
+    fn save_winner(gameID: u256, winner: Winners){
+        store(sha256(("winner", gameID)),winner);
+    }
+
+    fn get_winner(gameID: u256) -> Winners {
+        get(sha256(("winner", gameID, winner)));
+    }
+
+    fn calculate_winner(game: Game) -> Winners{
```
Comment (Braqzen) [inline Contracts/TicTacToe/src/main.sw:146]: pretty sure this function doesn't need to exist at all

File: Contracts/TicTacToe/src/main.sw [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+        ) {
+            return get_player_position_filled(gameID, 1);
+        }
+        elif (
+            get_player_position_filled(gameID, 3) == get_player_position_filled(gameID, 5) && get_player_position_filled(gameID, 5)  == get_player_position_filled(gameID, 7) && get_player_position_filled(gameID, 3) != Players.None
+        ) {
+            return get_player_position_filled(gameID, 2);
+        }
+        return Player.None;
+    }
```
Comment (Braqzen) [inline Contracts/TicTacToe/src/main.sw:240]: this is way too difficult to read. these can probably be deleted entirely if not rewritten to get a simple calculation to figure out the position

File: Contracts/TicTacToe/src/main.sw [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+
+    fn is_board_full(gameID: u256) -> bool{
```
Comment (Braqzen) [inline Contracts/TicTacToe/src/main.sw:242]: I don't know what the point of this function is

File: Contracts/TicTacToe/src/main.sw [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+        for (u8 x = 0; x < 9; x++) {
+            if (get_player_position_filled(gameID, x) == None) {
+                return false;
+            }
+        }
+        return true;
+    }
+}
```
Comment (Braqzen) [inline Contracts/TicTacToe/src/main.sw:250]:
```suggestion
}

```

File: Contracts/TicTacToe/tests/harness.rs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+    (instance, id)
+}
+
+#[tokio::test]
+async fn can_get_contract_id() {
+    let (_instance, _id) = get_contract_instance().await;
+
+    // Now you have an instance of your contract you can use to test each function
+}
```
Comment (Braqzen) [inline Contracts/TicTacToe/tests/harness.rs:27]:
```suggestion
}

```

File: Contracts/TicTacToe/tests/harness.rs [hunk 6]
```diff
@@ -0,0 +1,27 @@
+use fuel_tx::ContractId;
```
Comment (Braqzen) [inline Contracts/TicTacToe/tests/harness.rs:1]: update to latest release and write tests after logic is corrected

**Review by Braqzen — CHANGES_REQUESTED** [review]
The path should be `sway-applications/games/<name of game>/<dir containing sway code>`

You need to 

1. use `forc build` to check if your code even compiles
2. fix whatever errors it produces
3. use `forc fmt` to format your code when making a PR
4. add this to CI so the code is validated (sway-applications/.github/workflows/ci.yml , there's also a build.sh script around there)

Reconsider your data structures to allow for simpler data queries and general data handling.
Add error handling too.

After changes are made I'll add more in the next review

**Comment by adlerjohn** [comment]
Converting to draft just so it doesn't get merged by accident, until the target branch the brought up to date at least.

**Pull request closed without merging**